### PR TITLE
Implement BiCGSTAB

### PR DESCRIPTION
**Pull request by jlogan03**
Some thoughts about this implementation -

Pros
* It runs and is even fairly correct and a bit documented
* Implemented on f32 and f64 with generic indices

Cons
* Doesn't represent any particular reusable abstract pattern of iterative solver interface, and only makes a little progress toward the goals of #339 
* Solver call signature requires the turbof***
  * I haven't found any way around this - even when it's obvious from the inputs and outputs which variant should be selected, it still fails to select the correct variant without explicit type annotation
* Macros make things worse for the users in terms of readability, understanding errors, etc
* Not fully generic on value type
* Not generic on sparse vs. dense RHS and initial guess
* Can't be implemented directly on Complex<T> types
* Allocates then drops several vectors on each iteration which have the same sparsity at every iteration and could reuse the same memory if we had a mechanism for this
  * In practice, this isn't as bad as it sounds since the first few iterations will clear out a correctly sized region of heap s.t. the subsequent allocations are cheap to perform... as long as there is not much else happening on the machine

I spent a solid 8 hours attempting to make a version that is generic over the value type. This did not go well. 

Some of the implementations of the underlying matrix-vector and vector-scalar operations are not fully generic, so the trait bounds on the impl end up being more than 10 lines long and incredibly unclear as to which types will and will not be compatible, and the type system ultimately fails on overflow due to inferring generic outputs across iterations in the solver loop. 

This could all be fixed to some extent, but would require quite a bit of work to go back into the core of the crate and make a single definition of what trait bounds constitute a mathematical field over which we can do linear algebra, then propagate that definition to every struct and impl in the crate, which I believe is out of scope for this PR.

### CODE REVIEW

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+    rhat: CsVecI<T, I>, // Arbitrary w/ dot(rhat, r) != 0
+    p: CsVecI<T, I>,
+    // Intermediate scalars
+    rho: T,
+}
+
+macro_rules! bicgstab_impl {
+    ($T: ty) => {
+        impl<'a, I: SpIndex, Iptr: SpIndex> BiCGSTAB<'a, $T, I, Iptr> {
+            /// Initialize the solver with a fresh error estimat
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:117]:
```suggestion
            /// Initialize the solver with a fresh error estimate
```

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        println!("Hard restart count {:?}", res.hard_restart_count());
+
+        // Make sure the solved values match expectation
+        for (input, output) in
+            b.to_dense().iter().zip(b_recovered.to_dense().iter())
+        {
+            assert!(
+                (1.0 - input / output).abs() < tol,
+                "Solved output did not match input"
+            );
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:348]:
```suggestion
            assert!(
                approx::relative_eq!(input, output, max_relative = tol),
                "Solved output did not match input"
            );

```

Comment (mulimoen) [reply sprs/src/sparse/linalg/bicgstab.rs:348]: And `cargo add --dev approx` to add it to `Cargo.toml`

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+        println!("Iteration count {:?}", res.i());
+        println!("Soft restart count {:?}", res.soft_restart_count());
+        println!("Hard restart count {:?}", res.hard_restart_count());
+
+        // Make sure the solved values match expectation
+        for (input, output) in
+            b.to_dense().iter().zip(b_recovered.to_dense().iter())
+        {
+            assert!(
+                (1.0 - input / output).abs() < tol,
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:386]:
```suggestion
                approx::relative_eq!(input, output, max_relative = tol),
```

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            }
+
+            /// Set the minimum value of `rho` to trigger a soft restart
+            pub fn with_restart_threshold(mut self, thresh: $T) -> Self {
+                self.soft_restart_threshold = thresh;
+                self
+            }
+
+            /// Iteration number
+            pub fn i(&self) -> usize {
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:238]: Should this be given a more descriptive name such as `iteration_number`?

Comment (mulimoen) [reply sprs/src/sparse/linalg/bicgstab.rs:238]: But I have not read the papers and I don't know if this is standard lingo

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 5]
```diff
@@ -0,0 +1,391 @@
+//! Stabilized bi-conjugate gradient solver. Suitable for non-symmetric matrices.
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:1]:
```suggestion
//! Stabilized bi-conjugate gradient solver for solving Ax = b with x unknown. Suitable for non-symmetric matrices.
```

File: sprs/src/sparse/linalg/bicgstab.rs [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+//! println!("Iteration count {:?}", res.i());
+//! println!("Soft restart count {:?}", res.soft_restart_count());
+//! println!("Hard restart count {:?}", res.hard_restart_count());
+//!
+//! // Make sure the solved values match expectation
+//! for (input, output) in
+//!     b.to_dense().iter().zip(b_recovered.to_dense().iter())
+//! {
+//!     assert!(
+//!         (1.0 - input / output).abs() < tol,
```
Comment (mulimoen) [inline sprs/src/sparse/linalg/bicgstab.rs:64]:
```suggestion
//!         approx::relative_eq!(input, output, max_relative = tol
```

**Review by mulimoen — APPROVED** [review]
This is some fantastic work and it's great to see a proper general solver. This will be immensely useful to users.

Making everything generic has many drawbacks, the ergonomics can get much worse and it takes a lot longer to write for dubious benefit, the macro is great for implementing the two most useful impls and we can deal with `Complex` if anybody asks for it.

I only have some minor nits on this PR, I consider it good for a merge in the current state

File: sprs/src/sparse/linalg/bicgstab.rs (hunk 4 above)
Comment (jlogan03) [reply sprs/src/sparse/linalg/bicgstab.rs]: It's not standard lingo, just me being lazy - updated it to `iteration_count` for consistency with the other `*_count` values stored by the solver

**Comment by jlogan03** [comment]
@mulimoen I like the approx usage for testing, but had to roll it back because we don't keep approx as a default feature so it crashes the tests. I'd be happy to bring it back if we want to update either default features to include approx, or to include the approx feature for testing

**Comment by mulimoen** [comment]
I don't feel strongly about approx, bit of a pain to enable it for doctests. A bit of resolver history: The resolver previously merged features across build types, so having approx on during tests would enable it also for regular builds. Since we are using resolver v2 we can now use it for tests only

**Pull request merged**